design/authenticators/authn-k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,195 @@
+
+This option is not fully backwards-compatible and is quite complex so I wouldn't suggest to implement it.
+
+#### Decision
+
+Will be decided upon investigation.
+
+### Inject Client Cert response code
+Another aspect of the solution that we need to decide upon is the response code  that the request will return to the authn-client. Currently, we return a 200 OK response to indicate that the cert was installed without failure. This may be a wrong approach as this means that the server needs to wait until KubeCtl finished to install the certificate and we can let the authn-client hang until timeout in case of a load of client requests.
+
+We should consider another approach where we will return a 202 Accepted response once we get the request and run the kubectl exec flow in a separate thread. 
+
+To conclude the options, let's look at the following table:
+
+| **Response Code** | **Pros**                                                                                                                                   |
+|-------------------|--------------------------------------------------------------------------------------------------------------------------------------------|
+| 200 OK            | Full indication of the injection status for the client. 200 states that the cert is injected, and 401 states that the cert is not injected |
+| 202 Accepted      | Fast response so the client is not hanging and doesn't get to its timeout in case the injection takes too long                             |
+
+It's worth mentioning that we don't really have a backwards-compatibility issue as also today we return a 200 OK response before the cert is fully injected. Even before the wait that we added to client, it read the cert and failed - before retrying. So  changing the response to 202 Accepted is backwards-compatible.
+
+#### Investigation
+1. Is there a risk in waiting for the cert to be injected before returning a response?
+1. If we choose to return 202 Accepted, will we tell the client that the injection succeeded/failed? If so, how?
+
+#### Decision
+
+ Will be decided upon investigation.
+
+### Design
+[//]: # "Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about the expected user experience for the feature"
+TODO: Add diagram once we decide on a solution
+
+### Backwards compatibility
+[//]: # "Address how you are going to handle backwards compatibility, if necessary"
+At current, the authn-client sends the "inject-client-cert" request and then  tries to read the cert. As this solution will just fix the way that we inject the  cert, we should have any backwards-compatibility issues.
```

Comment:
```suggestion
At current, the authn-client sends the "inject-client-cert" request and then  tries to read the cert. As this solution will just fix the way that we inject the  cert, we shouldn't have any backwards-compatibility issues.
```

Reply:
Also, what about forward compatibility? In option 3, if a user takes the latest authn client that expects either 202 or 200 but also uses an older Conjur version, the new authn client would always get 200 and expect the file to exists. Right?

Reply:
yes, option 3 has backwards-compatibility issues. we will not go with it.